Add node shutdown KEP

bobbypage opened the pull request:
Enhancement issue: https://github.com/kubernetes/enhancements/issues/2000

karan commented:
/cc karan

karan commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
Shouldn't we then name it for example `00-kubelet.conf`? IOW, if the kubelet is expected to manage pod/machine shutdown, then it seems reasonable for the kubelet to (best-effort) take precendence.

bobbypage replied:
That's a good point -- the name of the file can be set to ensure kubelet has higher chance of being selected. I'll take a closer look at the best naming here so the config has a higher chance of being selected.

karan commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
"at least"

bobbypage replied:
Reworded.

karan commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
I think you want to move/add details about kubelet config API changes here.

bobbypage replied:
Good point, added some more details here.

karan commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
What cloud platforms are you going to test these on?

What systemd versions are you going to test on (likely only the ones provided by base images)?

bobbypage replied:
> What cloud platforms are you going to test these on?

Since the e2e tests only currently run on GCE, at least for e2e tests will be limited to GCP. We can run some manual tests on other cloud providers as a sanity check though.

> What systemd versions are you going to test on (likely only the ones provided by base images)?

As you mentioned, it'll be the same versions as provided in the base images the tests is running on.

rata commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,843 @@
+Locks".](https://www.freedesktop.org/wiki/Software/systemd/inhibit/)
+Applications can request to delay shutdown by taking an inhibitor lock by
+sending messages to logind over dbus. Applications can request up to
+`InhibitDelayMaxSec` (a setting configured in logind.conf`) for delay based
```
Niit-pick: I think there is a missing ` before logind.conf that makes the formatting wrong (because there is one at the end of logind.conf, closing)

bobbypage replied:
Thanks, fixed.

rata commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```suggestion
*   Kubelet is unable to update `InhibitDelayMaxSec` in logind to match that of
```

bobbypage replied:
Fixed.

SergeyKanzhelev commented on keps/sig-node/2000-graceful-node-shutdown/kep.yaml:
```diff
@@ -0,0 +1,34 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.19"
```
should it be 1.20? (based on field description)

bobbypage replied:
Yup, you're correct, fixed.

SergeyKanzhelev commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,843 @@
+Kubelet will propagate this signal to pods ensuring they can shutdown as
+gracefully as possible.
+
+### Goals
```
may be worth mentioning for scoping - does this KEP tries to address the expedited re-scheduling of pods or notification of API server that node is about to shutdown?

SergeyKanzhelev replied:
btw, there is another thing about shutdown: https://github.com/kubernetes/kubernetes/issues/58635 . That is resulting in this KEP: https://github.com/kubernetes/enhancements/issues/551. Just for the reference on other issues in this area. Just found out about it

bobbypage replied:
Thanks Sergey for pointing that out. Doesn't look like there is any activity on that KEP from October 2019, so will resolve this for now.

SergeyKanzhelev commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,843 @@
+
+Upon a shutdown occurring, Kubelet will gracefully terminate all the pods
+running on the node and taint itself with a new taint,
+`node.kubernetes.io/unschedulable` to avoid new pods being scheduled.
```
will this taint removed when machine is back up after the restart?

bobbypage replied:
That's the goal, ideally kubelet can update the status of node and add this taint when the shutdown starts. If the node is rebooted or another node comes up with the same name, ideally the node will update the status without this taint. Still a bit TBD, figuring out how this will look like implementation wise.

SergeyKanzhelev commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
so the test would run on kind? Or actual nodes would shutdown?

bobbypage replied:
I think all of the node e2e tests run directly on GCE currently, I'm not aware if kind is used for e2e tests. It'll be a bit challenging to develop e2e tests here as for a full test, I'm guessing we'll need to perform an actual shutdown which may interfere with the test runner. Need to look into this aspect further.

derekwaynecarr commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,843 @@
+
+Upon shutdown Kubelet will:
+
+1. Gracefully terminate all non critical system pods with a gracePeriodOverride
```
do we need a buffer to allow the kubelet to report back to the apiserver?

bobbypage replied:
to clarify, are you proposing before pods are terminated we first report status to apiserver, and only then start terminating pods?

dchen1107 replied:
I also hope we can have the status back to the apiserver, but hope the status include all pods are terminated or being terminated. Not sure if we can guarantee it though.

bobbypage replied:
I updated this to include that the upon shutdown, kubelet should also update it's Ready status to false with the reason `Node is shutting down`. 

Regarding the pod status being reported back as well, I think this would happen as part of the regular sync loop, but it might be difficult to guarantee depending on the timing as Dawn mentioned.

derekwaynecarr commented:
@bobbypage @mrunalp thanks for putting this together!

I have no major issues with the proposal as-is other than mechanical questions about testing.

I am happy if we want to merge and iterate on just that section of the KEP as the other parts of the KEP all lgtm.

I will let @dchen1107 take a final pass.

/assign @dchen1107

bobbypage commented:
Thanks @derekwaynecarr for taking a look! Will followup with @dchen1107 for a final pass. I put this KEP on agenda for upcoming SIG-Node, me and @mrunalp are happy to discuss this in more detail there.

dchen1107 commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,864 @@
+perform some series of actions before the shutdown completes through a
+mechanism called ["Inhibitor
+Locks".](https://www.freedesktop.org/wiki/Software/systemd/inhibit/)
+Applications can request to delay shutdown by taking an inhibitor lock by
```
nit: we might need to figure out the minimal systemd version for a working systemd inhibitor later.

bobbypage replied:
Updated with minimal systemd version:

> Inhibitor Locks were introduced to [systemd 183](https://lwn.net/Articles/499480/) (released in 2012).

dchen1107 commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,864 @@
+
+*   if the setting is greater than 0 seconds
+    *   kubelet will check with dbus current `InhibitDelayMaxSec` to check if
+        `kubeletConfig.ShutdownGracePeriod &lt;=` `InhibitDelayMaxSec`.
```
nit: <=

bobbypage replied:
Fixed.

dchen1107 commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,864 @@
+        writing a config file to `/etc/systemd/logind.conf.d/kubelet.conf`, and
+        sending a SIGHUP to logind to update the config setting to ensure that
+        the ShutdownGracePeriod from kubelet config is equal to
+        `InhibitDelayMaxSec`.
```
Reconfiguring the node through kubelet config is ok with me, but not sure if there are concerns from the community. It is worth mentioning at SIG Node meeting.

bobbypage replied:
We discussed this in SIG-Node today, I didn't hear any strong concerns from the community, so I think this should be fine for now.

dchen1107 commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,864 @@
+1. Gracefully terminate all non critical system pods with a gracePeriodOverride
+computed as `min(podSpec.terminationGracePeriodSeconds, ShutdownGracePeriod)`
+2. Gracefully terminate all critical system pods with gracePeriodOverride of 2
+seconds
```
Are we sure 2 seconds are enough here?

dchen1107 replied:
I saw you mentioned 2 second below, but since we are redesign the shutdown logic here, want to make sure that is enough for the critical system pods especially more and more system pods started to checkpoint.

bobbypage replied:
We discussed this during SIG-Node, the feedback sounded like 2 seconds may or may not work for various use cases, so I think it would makes sense to make this configurable via KubeletConfig. 

Unfortunately, this would result in two fields that need to be exposed in Kubelet Config: 

1. total shutdown grace period duration, `ShutdownGracePeriod`
2. shutdown grace period duration for critical pods, `ShutDownGracePeriodCritical` 

with the assumption that `ShutDownGracePeriodCritical` <  `ShutdownGracePeriod`

However, it sounds like this is worth doing, so I will update the KEP with this.

bobbypage replied:
Updated the KEP and changed from suggesting 2 seconds to instead provide a configurable duration via a new KubeletConfig option, `ShutdownGracePeriodCriticalPods`.

dchen1107 commented:
Overall looks good to me with some nits and questions.

marosset commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,864 @@
+aware of the pods and containers running on the machine and systemd will simply
+kill them as regular linux processes.
+
+### Background on Inhibitors
```
Windows has similar functionality to inhibitor locks.
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-shutdownblockreasoncreate?redirectedfrom=MSDN is the API call to delay a shutdown.

Can we make the specifics of how kubelet controls the shutdown delay configurable for different platforms?

marosset replied:
I can create a Background on Windows Shutdown section for inclusion in the KEP if that would be useful.

bobbypage replied:
Thanks @marosset, it would be great to add a Windows specific section, thanks for offering your help!

I'm thinking it would be good to merge the initial version of the KEP and then you can do a followup PR to add the windows specific info, WDYT?

marosset replied:
Sure, I can make a followup PR with the windows specific issue.

bobbypage replied:
Thanks Mark, will ping you after to get the windows specific info added.

dchen1107 commented:
We discussed this KEP at today's SIG node meeting. A couple more feedback raised at the meeting:

* Suggested by @marosset: Windows has similar APIs to register for shutdown events and delay shutdown. I'd love to see Windows support added here as well but not sure we'll have resources to work on this at the same time as the Linux implementations. I'll add links to the windows APIs to the doc PR tho.

We discussed it briefly, and we should include Windows Node support in KEP, but not alpha blocker here.

* Suggested by me in addition to two issues I raised above (1) 2s for critical system pods 2) reset node the config through kubelet's config): Instead of introducing the new taint, and new node condition, Kubelet could send node status by marking the node NotReady with detail message like: the node is shutting down.

dchen1107 commented:
@bobbypage I approved your KEP for now. Please address all above comments and ping me for another review.

/approve

andrewsykim commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
Does "Preemptible VMs" in all cloud providers trigger the shutdown event on termination?

andrewsykim replied:
(assuming the OS for the VM is running a compatible systemd version)

bobbypage replied:
That's a good question -- every cloud provider will eventually have to shutdown the VM and terminate it, so at some point the VM shutdown should be sent. 

For example, on GCE when a GCE Preemptible VM is terminated it gets 30 second time to shutdown, and the shutdown event is delivery at t-30 where t is when it will be forced shutdown. On AWS spot instances, the period is 2 minutes, but I'm unclear if the shutdown event is delivered a t-2min or at t itself. In addition to the systemd shutdown event, each cloud provider usually has a specific metadata server local to the VM that can be used to poll for preemption events. 

If the specific cloud provider doesn't actually trigger shutdown prior to the VM getting preempted, one workaround would be to have a cloud specific daemonset deployed that can poll the metadata server of the VM for preemption event and upon receiving the terminating event, simply trigger a node shutdown, i.e. `systemctl poweroff`, which will initiate kubelet graceful shutdown.

bobbypage commented:
Thanks @dchen1107 and @derekwaynecarr and rest of the folks from SIG-Node today for your feedback on this proposal. I will followup to address the comments here and would like to merge this and mark it implementable, so we can get the KEP in before the enhancements freeze for 1.20 which is October 6.

kikisdeliveryservice reviewed (commented):
> I will followup to address the comments here and would like to merge this and mark it implementable, so we can get the KEP in before the enhancements freeze for 1.20 which is October 6.

Hi there, If you want this in 1.20, you need to: Update the related Issue to get it in the milestone, add graduation criteria (alpha, beta, etc) and mark this as implementable.

rata commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,866 @@
+Kubelet will use the same existing
+[killPod](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/pod_workers.go#L292)
+function to perform the termination of pods, using `gracePeriodOverride` to set
+the appropriate grace period. During the termination process, normal [pod
```
Hi! Rodrigo here, we discussed the coordination with sidecar KEP. How are you doing? :)

As part of this KEP or sidecar we need to make a change to honor `gracePeriodOverride` ([current code][link] waits for pod.TermiantionGracePeriod to run the pre-stop hooks). Can you do it yourself (should be like 1 line change :))? It is inlikely that the sidecar KEP will be ready for 1.20, I can help if you want, though :)


[link]: https://github.com/kubernetes/kubernetes/blob/1b4fa94c2131dd0b11b470a280b967b0b10ffe2d/pkg/kubelet/kuberuntime/kuberuntime_container.go#L604-L626

bobbypage replied:
Good to hear from you Rodrigo, hope you're doing well!

Thanks for in pointing that, out added a note to KEP, looks like https://github.com/kubernetes/kubernetes/issues/92432 is the corresponding issue. I think it makes to change it as part of this KEP, since this it'll impact the behavior here and for sidecar proposal.

rata commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,866 @@
+[killPod](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/pod_workers.go#L292)
+function to perform the termination of pods, using `gracePeriodOverride` to set
+the appropriate grace period. During the termination process, normal [pod
+termination](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)
```
The `killContainer()` function (called from `killPod()`) may take longer than `gracePeriodOverride` if the pre-stop hooks take too long (see [here][link]). It may take 2 seconds more (without much precision, it only takes into account the waiting time, not the go-code executed. Is probably negligible for more use cases, not sure if some exotic use cases, like running on 5G on a train or those need more precision).

I guess this is completely fine for alpha and if during testing we detect several environments not finishing all pods correctly then we can see how to improve it. WDYT?


[link]: https://github.com/kubernetes/kubernetes/blob/1b4fa94c2131dd0b11b470a280b967b0b10ffe2d/pkg/kubelet/kuberuntime/kuberuntime_container.go#L627-L630

bobbypage replied:
Thanks for pointing that out, that's a good point. I agree, I don't think it should cause any major issue if the gracePeriod ends up being +- 2 seconds, but makes sense to reevaluate post alpha as you mentioned.

rata commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
I posted some comments on some edge cases where it might be possible to not properly kill all pods before shutdown. So, it can have similar effect to this risk, that's why I'm commenting here.

Maybe something worth doing to measure if we are effectively shutting down the pods before shutdown is to log after all pods were terminated (with fsync or some guarantees for it to be written to disk?), so if the log entry is missing we know something is not working as expected.

I think a simple log line might do the trick and help with bug reports (I'd not do anything more complicated unless really :)).

What do you think?

bobbypage replied:
That's a good point, in my testing I saw that I also had the logs I could fetch after shutdown and could see if the pods existing successfully, so I'm not sure if explicit `fsync` is needed. But I like your idea of adding a final log statement, like "ShutdownComplete", or something along those lines so it's possible to verify if the shutdown process from kubelet perspective was completed successfully.

bobbypage replied:
Added a note to the KEP about that in this section.

rata reviewed (commented):
@bobbypage Thanks for the KEP, LGTM. Added some simple comments and suggestions, but nothing major :)

bobbypage commented:
/retest

bobbypage commented:
I've updated the KEP based on the feedback so far (changed to use ReadyStatus and option to configure grace period critical pods) as mentioned in https://github.com/kubernetes/enhancements/pull/2001#issuecomment-701097113 . 

I've also updated the KEP and corresponding enhancement issue (https://github.com/kubernetes/enhancements/issues/2000) to implementable status targeting 1.20 as discussed during the SIG-Node meeting. Please let me know if there's any other concerns. 

Pinging @dchen1107 for final approval.

kikisdeliveryservice commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
Need to add something here for alpha in 1.20

bobbypage replied:
Added.

kikisdeliveryservice commented on keps/sig-node/2000-graceful-node-shutdown/kep.yaml:
```diff
@@ -0,0 +1,34 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.21"
```
v1.20 ^^^

bobbypage replied:
Updated.

kikisdeliveryservice reviewed (commented):
Noted 2 things from an enhancements team POV

bobbypage commented:
/retest

dchen1107 commented:
@bobbypage thanks for addressing our comments except Windows specific ones. Mark agreed to send the followup PRs to the KEP later.

/lgtm
/approve

smarterclayton commented on keps/sig-node/2000-graceful-node-shutdown/README.md:
```diff
@@ -0,0 +1,895 @@
+“critical system pods”, and regular pods. Critical system pods should be
+terminated last, because for example, if the logging pod is terminated first,
+logs from the other workloads will not be captured. Critical system pods are
+identified as those that are in the `system-cluster-critical` or
```
I missed this when it came out, but this is super concerning to me.  I don't think these priority classes are "special" in that the Kubelet should hardcode their use.  By hardcoding these, we FORCE workloads that interact with the kubelet or other system infra pods to be in these two priority classes, which breaks the orthogonality of scheduling and resource behavior from the kubelet.

Either there needs to be something that selects for priority classes to treat "specially", or these need to be configurable at startup time in the kubelet.  The former is more flexible, the latter may be more acceptable, but would not allow a service provider to allow users to make that orthogonal.

smarterclayton replied:
@bobbypage while reviewing the KEP (i was reading through it and thinking about the implications of the intersection with grace period when i noticed this), I think this has to be addressed before we go to beta.

bobbypage replied:
Hi @smarterclayton 

Thanks so much for providing your feedback and comments. 

@mrunalp and I discussed this topic in length as part of KEP design and we decided to use priority class of  `system-cluster-critical` and `system-node-critical` to separate "core system workloads (e.g logging deamonsets, etc)" vs regular pods and use that information to determine shutdown ordering.

Unfortunately as I'm sure you're aware there is no existing declarative mechanism to describe pod shutdown ordering and as such we decided to use pod priority as a signal instead.  This is similar to for example pod admission and OOM score adjustment which also use `IsCriticalPod()` as a signal.

@mrunalp and I definitely agree this is not perfect, and happy to discuss if you have some alternative ideas on how to improve it and a potentially better signal for us to use to determine pod shutdown order logic instead for beta moving forward.

bobbypage commented:
We had a chat today with @smarterclayton @mrunalp @SergeyKanzhelev regarding some of questions in https://github.com/kubernetes/enhancements/pull/2001#discussion_r565507971 

The main item we discussed was around the current design of having two shutdown stages, first being shutting down user workloads followed by "critical node system workloads" and the current pattern of using `system-cluster-critical` and `system-node-critical` to separate system workloads vs user workloads.

Some notes from our discussion:

* Currently users who want to get their workload into the later "critical" shutdown phase need to configure their pod spec's priority class to be `system-cluster-critical` or `system-node-critical`. However, some pods may want to get into "critical" shutdown phase, but are using some other priority class.
   * Does it make sense to make configurable what priority classes are considered critical vs not instead of hardcoding  `system-cluster-critical` and `system-node-critical`  as today?

* Node Graceful shutdown currently splits the total shutdown time into two phases, first user workloads, followed by critical pods.
   * Does it make sense to provide users ability to partition the shutdown time into more than the existing two phases?  
       * Perhaps providing ability to configure a list of priority classes (or maybe range of priorities) to a shutdown phase and an associated amount of time for each phase?

SergeyKanzhelev commented:
I think one more small piece of feedback was to allow to use leftover of one phase to extend another. I.e. if user pods terminated very quickly, let critical pods to use the rest of the time.

bobbypage commented:
To circle back on https://github.com/kubernetes/enhancements/pull/2001#issuecomment-770072076 regarding supporting "custom" priority classes for node shutdown other than `system-cluster-critical` and `system-node-critical`, we discussed this point in SIG-node with some other folks and @mrunalp. 

Ultimately we decided that it's not clear how many users are actually making use of custom priority classes and would want to partition the shutdown time per specific priority class, making it a bit of niche requirement and more complicated. 

We decided to proceed to beta with the current design. If we'll get more feedback / data that supporting configurable shutdown time per "custom" priority class, we can always add this capability as followup post beta.
